frontend/src/Frontend/UI/Widgets.hs

```diff
@@ -176,59 +177,102 @@ uiRealInputElement cfg = do
+
+    attemptCorrection :: Event t Text -> Event t Text
+    attemptCorrection = intercept (const True)
+
+    forceCorrection :: Event t Text -> Event t Text
+    forceCorrection = intercept (isJust . snd)
+
+    -- type inference goes crazy if we inline this - ghc bug?
+    inp' = fmapMaybe id $ sanitization inp
+
+  pure (ie, (fmap . fmap) fst $ sanitization val, inp')
+
 -- | Decimal input to the given precision. Returns the element, the value, and
 -- the user input events
-uiRealWithPrecisionInputElement
+uiNonnegativeRealWithPrecisionInputElement
   :: forall t m er a. (DomBuilder t m, MonadFix m, MonadHold t m)
   => Word8
   -> (Decimal -> a)
   -> InputElementConfig er t (DomBuilderSpace m)
-  -> m (InputElement er (DomBuilderSpace m) t, (Dynamic t (Maybe a), Event t a))
-uiRealWithPrecisionInputElement prec fromDecimal cfg = do
+  -> m (InputElement er (DomBuilderSpace m) t, Dynamic t (Maybe a), Event t a)
+uiNonnegativeRealWithPrecisionInputElement prec fromDecimal cfg = do
   rec
-    r <- inputElement $ cfg
+    (ie, val, input) <- uiCorrectingInputElement parse sanitize tshow $ cfg
       & initialAttributes %~ addInputElementCls . addNoAutofillAttrs
         . (<> ("type" =: "number" <> "step" =: stepSize <> "min" =: stepSize))
-      & inputElementConfig_setValue %~ (\e -> leftmost [fmapMaybe parseAndRound e, rounded])
-    widgetHold_ blank (bool blank (elClass "span" "real-input__rounded" (text $ "Rounded to " <> (tshow prec) <> " places")) <$> eHasRounded)
-    let parsedValue = fmap fst . parseDecimal <$> value r
-        parsedInput = fmapMaybe parseDecimal (_inputElement_input r)
-        -- Trim the users input if we had to round it
-        rounded = fmap (showDecimal . fst) $ ffilter snd parsedInput
-        eHasRounded = fmap snd parsedInput
-  pure
-    ( r
-    , (fmap fromDecimal <$> parsedValue
-      , fromDecimal . fst <$> parsedInput
-      )
-    )
-  where
-    showDecimal :: Decimal -> Text
-    showDecimal = tshow
-
-    parseAndRound :: Text -> Maybe Text
-    parseAndRound t = showDecimal . fst <$> parseDecimal t
-
-    -- Returns the decimal and whether or not it needed rounding to 'prec'
-    parseDecimal :: Text -> Maybe (Decimal, Bool)
-    parseDecimal t = ffor (readMay $ T.unpack t) $ \decimal ->
-      (D.roundTo prec decimal, D.decimalPlaces decimal > prec)
+    widgetHold_ blank $ ffor (fmap snd input) $ traverse_ $
+      elClass "span" "real-input__rounded" . text
+  pure (ie, (fmap . fmap) fromDecimal val, fmap (fromDecimal . fst) input)
 
+  where
     stepSize = "0." <> T.replicate (fromIntegral prec - 1) "0" <> "1"
+    parse = tread
+
+    sanitize :: Decimal -> Maybe (Decimal, Text)
+    sanitize decimal = asum
+      [ (0, "Cannot be negative")
+        <$ guard (decimal < 0)
+      , (D.roundTo prec decimal, ("Rounded to " <> tshow prec <> " places"))
+        <$ guard (D.decimalPlaces decimal > prec)
+      ]
 
+-- TODO: correct floating point decimals
```

Comment:
`min`/`max` attributes only seem to be applied to arrow-key updates? I could certainly add a minus sign or just add digits to the number without them affecting anything